src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignerInfo.cs

```diff
@@ -108,6 +108,118 @@ public SignerInfoCollection CounterSignerInfos
 
         public Oid SignatureAlgorithm => new Oid(_signatureAlgorithm);
 
+        public void AddUnsignedAttribute(AsnEncodedData unsignedAttribute)
+        {
+            int myIdx = _document.SignerInfos.FindIndexForSigner(this);
+
+            if (myIdx < 0)
+            {
+                throw new CryptographicException(SR.Cryptography_Cms_SignerNotFound);
+            }
+
+            ref SignedDataAsn signedData = ref _document.GetRawData();
+            ref SignerInfoAsn mySigner = ref signedData.SignerInfos[myIdx];
+
+            int existingAttribute = mySigner.UnsignedAttributes == null ? -1 : FindAttributeIndexByOid(mySigner.UnsignedAttributes, unsignedAttribute.Oid);
+
+            if (existingAttribute == -1)
+            {
+                // create a new attribute
+                AttributeAsn newUnsignedAttr;
+                using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
+                {
+                    writer.PushSetOf();
+                    writer.WriteEncodedValue(unsignedAttribute.RawData);
+                    writer.PopSetOf();
+
+                    newUnsignedAttr = new AttributeAsn
+                    {
+                        AttrType = new Oid(unsignedAttribute.Oid),
+                        AttrValues = writer.Encode(),
+                    };
+                }
+
+                int newAttributeIdx;
+
+                if (mySigner.UnsignedAttributes == null)
+                {
+                    newAttributeIdx = 0;
+                    mySigner.UnsignedAttributes = new AttributeAsn[1];
+                }
+                else
+                {
+                    newAttributeIdx = mySigner.UnsignedAttributes.Length;
+                    Array.Resize(ref mySigner.UnsignedAttributes, newAttributeIdx + 1);
+                }
+
+                mySigner.UnsignedAttributes[newAttributeIdx] = newUnsignedAttr;
+            }
+            else
+            {
+                // merge with existing attribute
+                ref AttributeAsn modifiedAttr = ref mySigner.UnsignedAttributes[existingAttribute];
+
+                using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
```

Comment:
@bartonjs should we also use BER for writing new attribute in this case? Since we do not know what rules were used for AsnEncodedData BER is the safest bet - do I understand that correctly?

Reply:
M...aybe.  An attribute whose RawData is `new byte[] { 0x30, 0x80, 0x00, 0x00 }` should test whether we're currently OK with indefinite encoding from AsnEncodedData for signed or unsigned attributes.  If we already have a DER-only restriction in the existing API (and netfx) then it's fine to keep that restriction here.

OTOH, if the other APIs are more lax, we probably need to be more lax here.

Reply:
@bartonjs do we currently have any easy way to check if AsnEncodedData can use indefinite length encoding? Where do I check for such restriction? Will writer detect such mismatch and throw right exception?

Reply:
If the RawData value is `30 80 00 00` (constructed sequence, indefinite length, end of contents) and things work, we handle indefinite.

RawData being `30 01` should fail (constructed sequence, one byte (contents missing), as should `30 00 30 00` (two empty sequences).